lib/supavisor/db_handler.ex

```diff
@@ -255,4 +263,83 @@ defmodule Supavisor.DbHandler do
 
     :keep_state_and_data
   end
+
+  # Adapted from Postgrex.Protocol
+  defp handshake(sock, %{upstream_ssl: true} = auth), do: ssl(sock, auth)
+  defp handshake(sock, auth), do: send_startup(sock, auth)
+
+  defp ssl(sock, auth) do
+    case sock_send(sock, Server.ssl_request()) do
+      :ok -> ssl_recv(sock, auth)
+      {:disconnect, _, _} = dis -> dis
+    end
+  end
+
+  @spec ssl_recv(term(), map) ::
+          {:ok, {:ssl, :sslsock}}
+          | {:error, :ssl_not_available | :ssl_connect_error}
+  defp ssl_recv({:gen_tcp, sock} = s, auth) do
+    case :gen_tcp.recv(sock, 1, :infinity) do
+      {:ok, <<?S>>} ->
+        ssl_connect(s, auth)
+
+      {:ok, <<?N>>} ->
+        Logger.error("SSL requested but server says it's not available")
+        {:error, :ssl_not_available}
+
+      {:error, reason} ->
+        Logger.error("Error when receiving SSL response: #{inspect(reason)}")
+
+        {:error, :ssl_connect_error}
+    end
+  end
+
+  defp ssl_connect({:gen_tcp, sock}, auth, timeout \\ 5000) do
+    opts =
+      case auth.upstream_verify do
+        :peer ->
+          [
+            verify: :verify_peer,
+            cacerts: [auth.upstream_tls_ca],
+            customize_hostname_check: [{:match_fun, fn _, _ -> true end}]
+          ]
+
+        :none ->
+          [verify: :verify_none]
+      end
+
+    case :ssl.connect(sock, opts, timeout) do
+      {:ok, ssl_sock} ->
+        send_startup({:ssl, ssl_sock}, auth)
+
+      {:error, reason} ->
+        Logger.error("Error when connecting with SSL: #{inspect(reason)}")
+
+        {:error, :ssl_connect_error}
+    end
+  end
+
+  defp send_startup(sock, auth) do
+    msg =
+      :pgo_protocol.encode_startup_message([
+        {"user", auth.user},
+        {"database", auth.database},
+        {"application_name", auth.application_name}
+      ])
+
+    :ok = sock_send(sock, msg)
+    {:ok, sock}
+  end
+
+  defp sock_send({mod, sock}, data) do
+    mod.send(sock, data)
+  end
+
+  defp set_sock_opts({:gen_tcp, sock}, opts) do
+    :inet.setopts(sock, opts)
```

Comment:
we should check the return value, either here or at the call site:

```suggestion
    :ok = :inet.setopts(sock, opts)
```